### PR TITLE
amm: Replace getLPSharesAmounts with getTokensInGivenMaximum

### PR DESCRIPTION
**Pull request by axic**
This is based on AMMBalancesHelpers.getLPSharesAmounts

### CODE REVIEW

File: contracts/amm/TempusAMM.sol [hunk 1]
```diff
@@ -158,18 +158,6 @@ contract TempusAMM is ITempusAMM, ERC20, Pausable, Ownable {
         emit Join(amountToken0, amountToken1, lpTokensOut);
     }
 
-    function compositionBalanceOf(address account)
```
Comment (axic) [inline contracts/amm/TempusAMM.sol:161]: Moved this to be properly sorted, was badly placed in the merge.

File: contracts/amm/TempusAMM.sol [hunk 2]
```diff
@@ -461,6 +461,39 @@ contract TempusAMM is ITempusAMM, ERC20, Pausable, Ownable {
...
+        external
+        view
+        override
+        returns (uint256 token0Amount, uint256 token1Amount)
+    {
+        require(totalSupply() > 0, "not initialised");
+        uint256 token0Balance = token0.balanceOf(address(this));
+        uint256 token1Balance = token1.balanceOf(address(this));
+
+        (uint256 token0Rate, uint256 token1Rate) = (token0Balance > token1Balance)
```
Comment (axic) [inline contracts/amm/TempusAMM.sol:474]: @mijovic please double check this

Comment (mijovic) [reply contracts/amm/TempusAMM.sol:486]: This looks good

Comment (mijovic) [reply contracts/amm/TempusAMM.sol:486]: I would just rename `token0Rate` to maybe `token0Scale` or something like that. This way it is conflicting with actual tokenRates

File: contracts/amm/interfaces/ITempusAMM.sol [hunk 3]
```diff
@@ -152,6 +152,11 @@ interface ITempusAMM is IERC20, IRateProvider, IOwnable {
     /// @return amount of LP tokens that could be received
     function getLPTokensOutForTokensIn(uint256 token0AmountIn, uint256 token1AmountIn) external view returns (uint256);
 
+    function getLPSharesAmounts(uint256 sharesAmount)
```
Comment (mijovic) [inline contracts/amm/interfaces/ITempusAMM.sol:155]: Natspec

File: test/TempusController.test.ts [hunk 4]
```diff
@@ -131,20 +131,25 @@ describeForEachPool("TempusController", (testPool:PoolTestFixture) =>
...
 
-    it("verifies depositing ERC20 BT and providing liquidity to a non initialized AMM reverts", async () =>
+    it("verifies depositing (ERC20) BT and providing liquidity to a non initialized AMM reverts", async () =>
     {
-      const invalidAction = controller.depositAndProvideLiquidity(testPool, user1, 123, true);
-      (await expectRevert(invalidAction)).to.equal(":AMMNotInitializedYet");
+      const amount = testPool.type == PoolType.Lido ? 0 : 123;
+      const ethValue = testPool.type == PoolType.Lido ? 123 : 0;
+      const invalidAction = controller.depositAndProvideLiquidity(testPool, user1, amount, true, ethValue);
+      if (testPool.type != PoolType.Lido) {
```
Comment (axic) [inline test/TempusController.test.ts:142]: This should not be needed as the parameters should work, but for some reason it doesn't yet.

File: contracts/amm/TempusAMM.sol (hunk 2 above)
Comment (axic) [reply contracts/amm/TempusAMM.sol]: Done.

File: contracts/amm/TempusAMM.sol [hunk 5]
```diff
@@ -451,6 +453,24 @@ contract TempusAMM is ITempusAMM, ERC20, Pausable, Ownable {
                 );
     }
 
+    function getLPSharesAmounts(uint256 sharesAmount)
```
Comment (axic) [inline contracts/amm/TempusAMM.sol:456]: Need better naming for both the function and the argument.

Comment (axic) [reply contracts/amm/TempusAMM.sol:456]: How about:
```solidity
    /// @dev queries the amount tokens to deposit (out of a maximum) to maintain balance of AMM
    /// @param maxAmount maximum amount of tokens to deposit
    /// @return token0Amount actual amount of token0 to deposit
    /// @return token1Amount actual amount of token1 to deposit
    function getTokensInGivenMaximum(uint256 maxAmount)
        external
        view
        returns (uint256 token0Amount, uint256 token1Amount);
```

Comment (mijovic) [reply contracts/amm/TempusAMM.sol:456]: This sounds very good to me

File: test/TempusController.test.ts (hunk 4 above)
Comment (mijovic) [reply test/TempusController.test.ts:143]: pretty weird, I do not have ideas at the moment

File: contracts/amm/TempusAMM.sol (hunk 5 above)
Comment (axic) [reply contracts/amm/TempusAMM.sol]: Changed.

File: test/TempusController.test.ts [hunk 6]
```diff
@@ -131,13 +131,15 @@ describeForEachPool("TempusController", (testPool:PoolTestFixture) =>
...
     });
 
-    it("verifies depositing ERC20 BT and providing liquidity to a non initialized AMM reverts", async () =>
+    it("verifies depositing (ERC20) BT and providing liquidity to a non initialized AMM reverts", async () =>
     {
-      const invalidAction = controller.depositAndProvideLiquidity(testPool, user1, 123, true);
-      (await expectRevert(invalidAction)).to.equal(":AMMNotInitializedYet");
+      const amount = 123;
+      const ethValue = testPool.type == PoolType.Lido ? amount : 0;
+      const invalidAction = controller.depositAndProvideLiquidity(testPool, user1, amount, true, ethValue);
```
Comment (axic) [inline test/TempusController.test.ts:141]: This was actually "broken", but it worked because the check for "amm not initialised" took precedence.

**Comment by axic** [comment]
Gas before:
```
|  Contract             ·  Method                                ·  Min        ·  Max        ·  Avg           ·  # calls      ·  eur (avg)  │
|  TempusController     ·  depositAndProvideLiquidity            ·     223003  ·     327861  ·        252754  ·           36  ·          -  │
```

Gas after:
```
|  TempusController     ·  depositAndProvideLiquidity            ·     222382  ·     327240  ·        252084  ·           36  ·          -  │
```

It even saves a tiny amount of gas.

**Review by mijovic — APPROVED** [review]

**Pull request merged**